### PR TITLE
Fix/dapp crash on authenticate

### PR DESCRIPTION
**Pull request by alexjba**
### What does the PR do

Fixing #16032

There are two issues fixed by this PR:
1. Crashing on authentication while signing dapp messages. In this case the root cause was the usage of model data items stored in js. When the modal changes, the underlying data changes as well and the app would crash in the JS execution.

- The quick fix is to create deep copies for the model data to be stored in JS.

2. Fixing the desynchronisation of the dapp connection status. This issue had lots of apparently random symptoms that could have been wrongly attributed to the shared cookies!

-    The DAppsListProvider needs to receive all the user accounts so that it can process all dapps the user is connected to (the dapps filtering based on selected account is an internal impl. detail)
- Call `revokeSession` on disconnect only if the dapp cannot be found in the WC active sessions. There is some overlapping between Browser connector and Wallet connect because both operate on the same data stored in the DB. If Browser connector controller removes WC data, the WC disconnect flows cannot be successfully completed.
- Reuse the same notification flow for Browser connector as it's used for WC
- Fix dapp filtering when processing the dapps that will be displayed in the UI.

### Affected areas

Wallet connect
Browser connect plugin

### How to test

1. App desynchronisation:
        Connect multiple dapps to multiple accounts. Try to switch between accounts when connecting dapps and check for the connection status in the dapp list.
3. Crash
       Connect to a dapp and request a sign. Click sign a when the authentication dialog is prompted let the app sit in the background for a few minutes. After authenticating the app would crash.

### Risk 

It has an impact on disconnect feature for Browser plugin that I can't assess!

Tick **one**:
- [x] Low risk: 2 devs MUST perform testing as specified above and attach their results as comments to this PR **before** merging.
- [ ] High risk: QA team MUST perform additional testing in the specified affected areas **before** merging.

### CODE REVIEW

**Comment by caybro** [comment]
Legit `ui-test` failure there:
```
[2024-08-09T11:07:15.968Z] 3: FAIL!  : QmlTests::DAppsListProvider::test_TestUpdateDapps() expected iconUrl to be set
[2024-08-09T11:07:15.968Z] 3:    Actual   (): https://tst1.com
[2024-08-09T11:07:15.968Z] 3:    Expected (): https://tst1.com/u/1
[2024-08-09T11:07:15.968Z] 3:    Loc: [/home/jenkins/workspace/s_linux_x86_64_tests-ui_PR-16050/storybook/qmlTests/tests/tst_DAppsWorkflow.qml(606)]
```

**Review by caybro — APPROVED** [review]
LGTM, just the test fix plus the notification question

File: ui/app/AppLayouts/Wallet/services/dapps/DappsConnectorSDK.qml [hunk 1]
```diff
@@ -482,7 +508,6 @@ WalletConnectSDKBase {
             controller.recallDAppPermission(dAppUrl)
             const session = { url: dAppUrl, name: "", icon: "" }
             root.wcService.connectorDAppsProvider.revokeSession(JSON.stringify(session))
-            root.wcService.displayToastMessage(qsTr("Disconnected from %1").arg(dAppUrl), false)
```
[the remark this answers is not in the document]
Comment (alexjba) [reply ui/app/AppLayouts/Wallet/services/dapps/DappsConnectorSDK.qml:485]: Yes, the notification is handled in the service component.

File: ui/app/AppLayouts/Wallet/services/dapps/WalletConnectService.qml [hunk 2]
```diff
@@ -137,13 +138,19 @@ QObject {
...
                         (accountsInSession.includes(dappsProvider.selectedAddress)))
                     {
+                        dappFoundInWcSessions = true
                         wcSDK.disconnectSession(topic)
                     }
                 }
             }
-        });
 
-        root.revokeSession(url)
```
Comment (alexjba) [inline ui/app/AppLayouts/Wallet/services/dapps/WalletConnectService.qml:146]: @kounkou WDYT of this change? The intention here is not to delete the WC session from DB before trying to process it with the WC flows.
So the process is changed as follows:
- Try to find the session in the WC sessions.If the session is not found in the WC sessions then process it through the Browser connect impl

Comment (alexjba) [reply ui/app/AppLayouts/Wallet/services/dapps/WalletConnectService.qml:146]: I have no idea how to test the Browser connect though..Any help is much appreciated.

File: ui/app/AppLayouts/Wallet/services/dapps/DappsConnectorSDK.qml [hunk 3]
```diff
@@ -148,9 +148,22 @@ WalletConnectSDKBase {
...
             })
+
+            if (!account) {
+                return null
+            }
+
+            // deep copy to avoid operations on the original object
+            try {
+                return JSON.parse(JSON.stringify(account))
+            }
```
Comment (kounkou) [inline ui/app/AppLayouts/Wallet/services/dapps/DappsConnectorSDK.qml:162]: nit: Should the catch go on the same line as `}` seems like you are consistent for other changes, so feel free to ignore.

**Review by kounkou — APPROVED** [review]
Amazing job, thanks!

**Comment by virginiabalducci** [comment]
Hi all, I'm currently testing these fixes. 

Fix 1 "Crashing on authentication while signing dapp messages" looks good, I haven't experienced any crash by following teh steps to try to reproduce the issue.

For issue 2, I'm still testing. 

While testing on the build from this PR, I'm noticing a weird behavior:

1. Have the browser connect extension installed and open the status build from this PR.
2. Go to any of these sites
https://yearn.fi/v3
https://pancakeswap.finance/

Actual behavior: 
The Connection request pop up on the Status desktop app opens by itself as soon as the site loads, without user initiating the connection by clicking on "Connect wallet"
If user tries to reject the connection, the pop up does not go away
The only way for the pop up to go away is to close the site
[console log browser connect pop up.txt](https://github.com/user-attachments/files/16566364/console.log.browser.connect.pop.up.txt)

https://github.com/user-attachments/assets/56adddad-1700-4522-8be6-2f439e211864

Expected
Need to confirm but I suppose the app should not connect by itself unless the user initiates the connection by clicking on the site's connect link.
If user chooses to reject the connection, the pop up should close.

I will continue testing issue 2 but wanted to bring this to your attention in the meantime, thanks!

**Comment by alexjba** [comment]
> Hi all, I'm currently testing these fixes. 
> 
> Fix 1 "Crashing on authentication while signing dapp messages" looks good, I haven't experienced any crash by following teh steps to try to reproduce the issue.
> 
> For issue 2, I'm still testing. 
> 
> While testing on the build from this PR, I'm noticing a weird behavior:
> 
> 1. Have the browser connect extension installed and open the status build from this PR.
> 2. Go to any of these sites
> https://yearn.fi/v3
> https://pancakeswap.finance/
> 
> Actual behavior: 
> The Connection request pop up on the Status desktop app opens by itself as soon as the site loads, without user initiating the connection by clicking on "Connect wallet"
> If user tries to reject the connection, the pop up does not go away
> The only way for the pop up to go away is to close the site
> [console log browser connect pop up.txt](https://github.com/user-attachments/files/16566364/console.log.browser.connect.pop.up.txt)
> 
> https://github.com/user-attachments/assets/56adddad-1700-4522-8be6-2f439e211864
> 
> Expected
> Need to confirm but I suppose the app should not connect by itself unless the user initiates the connection by clicking on the site's connect link.
> If user chooses to reject the connection, the pop up should close.
> 
> I will continue testing issue 2 but wanted to bring this to your attention in the meantime, thanks!

Thanks @virginiabalducci ! Looks like the request queue is not properly cleared. Will have a look!

**Comment by virginiabalducci** [comment]
I'm testing issue 2 with 2 different accounts, and I don't see the app being connected on the desktop app's end.
The Status wallet is connected on the browser's end, but is not present on the list of dApps from the desktop app, nor user is asked to agree to connect the dapp with a pop up on the desktop app.

I tested two scenarios: logging out of 1 account, then logging into a different account, no cookies cleared. Second test is the same, but after logging out of first user, did a clear cache and then logged into a different account.

https://github.com/user-attachments/assets/dc8d64da-34ab-4deb-b0d5-0378274a4b6b

[console log browser connect 2 accounts.txt](https://github.com/user-attachments/files/16567060/console.log.browser.connect.2.accounts.txt)
[console log browser connect 2 accounts clear cookies.txt](https://github.com/user-attachments/files/16567145/console.log.browser.connect.2.accounts.clear.cookies.txt)

https://github.com/user-attachments/assets/79233861-b270-4c35-889a-fbca063a414f

**Comment by kounkou** [comment]
> I'm testing issue 2 with 2 different accounts, and I don't see the app being connected on the desktop app's end. The Status wallet is connected on the browser's end, but is not present on the list of dApps from the desktop app, nor user is asked to agree to connect the dapp with a pop up on the desktop app.
> 

Hi @virginiabalducci,

Shouldn't you have the `Status` in connect Wallet ? See below screenshot

Screenshot from 2024-08-09 16-55-11

**Comment by alexjba** [comment]
> > Hi all, I'm currently testing these fixes.
> > Fix 1 "Crashing on authentication while signing dapp messages" looks good, I haven't experienced any crash by following teh steps to try to reproduce the issue.
> > For issue 2, I'm still testing.
> > While testing on the build from this PR, I'm noticing a weird behavior:
> > 
> > 1. Have the browser connect extension installed and open the status build from this PR.
> > 2. Go to any of these sites
> >    https://yearn.fi/v3
> >    https://pancakeswap.finance/
> > 
> > Actual behavior:
> > The Connection request pop up on the Status desktop app opens by itself as soon as the site loads, without user initiating the connection by clicking on "Connect wallet"
> > If user tries to reject the connection, the pop up does not go away
> > The only way for the pop up to go away is to close the site
> > [console log browser connect pop up.txt](https://github.com/user-attachments/files/16566364/console.log.browser.connect.pop.up.txt)
> > 
> >   
> >     
> >     
> > 
> >     Screen.Recording.2024-08-09.at.5.02.36.PM.mov
> >     
> >   
> > 
> >   
> > 
> >   
> > 
> > 
> > Expected
> > Need to confirm but I suppose the app should not connect by itself unless the user initiates the connection by clicking on the site's connect link.
> > If user chooses to reject the connection, the pop up should close.
> > I will continue testing issue 2 but wanted to bring this to your attention in the meantime, thanks!
> 
> Thanks @virginiabalducci ! Looks like the request queue is not properly cleared. Will have a look!

@virginiabalducci Looks like it's not related to this PR. Can be reproduced on the release branch as well.

@kounkou I've added some logs and it seems that we generate connection requests immediately after rejecting the request. Any idea what could be the trigger for these requests?

```
DBG 2024-08-12 11:44:20.328+03:00 qt message                                 topics="qt" tid=10942286 text="ConnectorSDK.onDappRequestsToConnect; requestId:  5b8c3ceb563f82d7f1574b3643e0396a009fead33bc833fcda42962a6d221c7f dappInfoString:  {\"icon\":\"https://yearn.fi/favicons/favicon-32x32.png\",\"name\":\"yearn.fi\",\"url\":\"https://yearn.fi\"}" file=qrc:/app/AppLayouts/Wallet/services/dapps/DappsConnectorSDK.qml:547 category=qml
WRN 2024-08-12 11:44:20.334+03:00 qt warning                                 topics="qt" tid=10942286 text="Warning: Multi-selection is disabled, but multiple items are selected. Automatically selecting the first inserted item." file=qrc:/app/AppLayouts/Wallet/views/NetworkSelectorView.qml:43 category=qml
WRN 2024-08-12 11:44:20.358+03:00 qt warning                                 topics="qt" tid=10942286 text="QML Binding: Binding loop detected for property \"value\"" file=qrc:/imports/shared/controls/AccountSelector.qml:136 category=default
DBG 2024-08-12 11:44:21.739+03:00 NewBE_callPrivateRPC                       topics="rpc" tid=10942286 file=core.nim:27 rpc_method=wakuext_slowdownArchivesImport
DBG 2024-08-12 11:44:22.865+03:00 qt message                                 topics="qt" tid=10942286 text="ConnectorSDK.rejectSession; requestId:  5b8c3ceb563f82d7f1574b3643e0396a009fead33bc833fcda42962a6d221c7f" file=qrc:/app/AppLayouts/Wallet/services/dapps/DappsConnectorSDK.qml:616 category=qml
DBG 2024-08-12 11:44:22.866+03:00 NewBE_callPrivateRPC                       topics="rpc" tid=10942286 file=core.nim:27 rpc_method=connector_requestAccountsRejected
DBG 2024-08-12 11:44:23.740+03:00 qt message                                 topics="qt" tid=10942286 text="ConnectorSDK.onDappRequestsToConnect; requestId:  1606dced0efc0fde2a9449a23fe087b9362f56b8029fca6e5338e14869381f14 dappInfoString:  {\"icon\":\"https://yearn.fi/favicons/favicon-32x32.png\",\"name\":\"yearn.fi\",\"url\":\"https://yearn.fi\"}" file=qrc:/app/AppLayouts/Wallet/services/dapps/DappsConnectorSDK.qml:547 category=qml
WRN 2024-08-12 11:44:23.744+03:00 qt warning                                 topics="qt" tid=10942286 text="Warning: Multi-selection is disabled, but multiple items are selected. Automatically selecting the first inserted item." file=qrc:/app/AppLayouts/Wallet/views/NetworkSelectorView.qml:43 category=qml
WRN 2024-08-12 11:44:23.761+03:00 qt warning                                 topics="qt" tid=10942286 text="QML Binding: Binding loop detected for property \"value\"" file=qrc:/imports/shared/controls/AccountSelector.qml:136 category=default
DBG 2024-08-12 11:44:24.514+03:00 qt message                                 topics="qt" tid=10942286 text="ConnectorSDK.rejectSession; requestId:  1606dced0efc0fde2a9449a23fe087b9362f56b8029fca6e5338e14869381f14" file=qrc:/app/AppLayouts/Wallet/services/dapps/DappsConnectorSDK.qml:616 category=qml
DBG 2024-08-12 11:44:24.515+03:00 NewBE_callPrivateRPC                       topics="rpc" tid=10942286 file=core.nim:27 rpc_method=connector_requestAccountsRejected
DBG 2024-08-12 11:44:24.742+03:00 qt message                                 topics="qt" tid=10942286 text="ConnectorSDK.onDappRequestsToConnect; requestId:  4206b661be262f35363c6d1c7b077f1d3d31ecf28c55ea54ed1bdd8c053861b2 dappInfoString:  {\"icon\":\"https://yearn.fi/favicons/favicon-32x32.png\",\"name\":\"yearn.fi\",\"url\":\"https://yearn.fi\"}" file=qrc:/app/AppLayouts/Wallet/services/dapps/DappsConnectorSDK.qml:547 category=qml
WRN 2024-08-12 11:44:24.744+03:00 qt warning                                 topics="qt" tid=10942286 text="Warning: Multi-selection is disabled, but multiple items are selected. Automatically selecting the first inserted item." file=qrc:/app/AppLayouts/Wallet/views/NetworkSelectorView.qml:43 category=qml
WRN 2024-08-12 11:44:24.756+03:00 qt warning                                 topics="qt" tid=10942286 text="QML Binding: Binding loop detected for property \"value\"" file=qrc:/imports/shared/controls/AccountSelector.qml:136 category=default
DBG 2024-08-12 11:44:25.784+03:00 NewBE_callPrivateRPC                       topics="rpc" tid=10942286 file=core.nim:27 rpc_method=wakuext_speedupArchivesImport
```

**Comment by alexjba** [comment]
> I'm testing issue 2 with 2 different accounts, and I don't see the app being connected on the desktop app's end. The Status wallet is connected on the browser's end, but is not present on the list of dApps from the desktop app, nor user is asked to agree to connect the dapp with a pop up on the desktop app.
> 
> I tested two scenarios: logging out of 1 account, then logging into a different account, no cookies cleared. Second test is the same, but after logging out of first user, did a clear cache and then logged into a different account.
> 
>  Screen.Recording.2024-08-09.at.6.28.31.PM.mov 
> 
> [console log browser connect 2 accounts.txt](https://github.com/user-attachments/files/16567060/console.log.browser.connect.2.accounts.txt) [console log browser connect 2 accounts clear cookies.txt](https://github.com/user-attachments/files/16567145/console.log.browser.connect.2.accounts.clear.cookies.txt)
> 
>  Screen.Recording.2024-08-09.at.6.36.50.PM.mov

@virginiabalducci This flow might not be implemented in the Browser plugin integration 🤔 From what I see the app we don't fetch the connected dapps from the backend after login.

My suggestion would be to add different tasks for both issues.

**Review by virginiabalducci — APPROVED** [review]
Looks good, I will check the two issues reported and if needed create separate bugs from them.

**Comment by alexjba** [comment]
@iurimatias @jrainville This one is ready for merge 🙏

**Comment by anastasiyaig** [comment]
@jrainville we need to cherry pick that pls and increase the RC number :D

**Pull request merged**